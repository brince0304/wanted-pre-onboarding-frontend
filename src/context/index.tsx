import React, {
    createContext, Dispatch,
    useContext, useEffect,
    useReducer,

} from "react";

type Action = { type: "SET_TOKEN"; token: string | null; };

type State = { accessToken: string | null; }

type AppDispatch = Dispatch<Action>;

const initialState: State = {
    accessToken: null,
}

const stateName = 'tokenState';
const TokenStateContext = createContext<State | null>(null);
const TokenDispatchContext = createContext<AppDispatch | null>(null);

const TokenReducer = (state: State, action: Action) => {
    switch (action.type) {
        case "SET_TOKEN":
            return {
                ...state,
                accessToken: action.token
            };
        default:
            throw new Error("Unhandled action");
    }
};

export function TokenProvider({children}: { children: React.ReactNode }) {

    const [state, dispatch] = useReducer(TokenReducer, initialState);

    useEffect(() => {
        // 로컬 스토리지에서 상태 값 불러오기
        const storedState = localStorage.getItem(stateName);
        if (storedState) {
            setToken(dispatch, JSON.parse(storedState).accessToken);
        }
    }, []);

    useEffect(() => {
        // 상태 값 로컬 스토리지에 저장하기
        localStorage.setItem(stateName, JSON.stringify(state));
    }, [state]);


    return (
        <TokenStateContext.Provider value={state}>
            <TokenDispatchContext.Provider value={dispatch}>
                {children}
            </TokenDispatchContext.Provider>
        </TokenStateContext.Provider>
    );
}

export function useTokenState() {
    const state = useContext(TokenStateContext);
    if (!state) {
        throw new Error("Connot find TokenStateProvider");
    }
    return state;
}

export function useTokenDispatch() {
    const dispatch = useContext(TokenDispatchContext);
    if (!dispatch) {
        throw new Error("Cannot find DispatchProvider");
    }
    return dispatch;
}

export const setToken = (dispatch: AppDispatch, token: string) => {
    dispatch({type: "SET_TOKEN", token});
}

export const getTokenFromLocalStorage = () => {
    const tokenState = localStorage.getItem(stateName);
    if (tokenState) {
        return JSON.parse(tokenState).accessToken;
    }
    return null;
}

export const removeTokenFromLocalStorage = () => {
    localStorage.removeItem(stateName);
}
