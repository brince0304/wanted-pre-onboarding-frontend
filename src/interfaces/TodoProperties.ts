export type TodoProperties = TodoPropertiesChild[];
export interface TodoPropertiesChild {
	id: number;
	todo: string;
	isCompleted: boolean;
	userId: number;
}