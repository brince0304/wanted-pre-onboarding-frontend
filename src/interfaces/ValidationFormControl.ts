import React from "react";

interface ValidationFormControl {
    testId: string,
        validationMessage: string,
        validationTrueMessage : string,
        placeholder: string,
        type: string,
        value: string,
        onChange: (e: React.ChangeEvent<HTMLInputElement>) => void,
        validation: boolean,
        label: string,
}

export default ValidationFormControl;
