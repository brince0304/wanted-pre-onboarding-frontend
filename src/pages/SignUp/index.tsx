import {Box, Button, Typography} from "@mui/material";
import ValidationInput from "../../components/ValidationInput";
import React, {FormEvent, useCallback, useEffect} from "react";
import {useFormControl} from "../../hooks/useFormControl";
import {useNavigate} from "react-router";
import {postSignup} from "../../apis";
import styled from "@emotion/styled";
import {isAuthorized} from "../../apis/utils/auth";

const StyledFormControl = styled.form`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  width: 100%;
  gap: 20px;
`;

const StyledInputBox = styled(Box)`
  width: 100%;
  height: 100%;
`;

const StyledSignupButton = styled(Button)`
  width: 100%;
  height: 100%;
`;

const SignUp = () => {
    const navigate = useNavigate();
    const emailRegex = /@/;
    const passwordRegex = /.{8,}/;
    const [onChangeEmail, email, emailValidation, setEmailValidation] = useFormControl({regex: emailRegex});
    const [onChangePassword, password, passwordValidation] = useFormControl({regex: passwordRegex});
    const [emailValidationMessage, setEmailValidationMessage] = React.useState<string>("이메일 형식이 아닙니다.");
    const [passwordValidationMessage, setPasswordValidationMessage] = React.useState<string>("비밀번호는 8자리 이상이어야 합니다.");

    const isAuth = isAuthorized();
    useEffect(() => {
        if (isAuth) {
            navigate("/todo");
        }
    });

    const emailInputProps = {
        testId: "email-input",
        validationMessage: emailValidationMessage,
        validationTrueMessage: "이메일 형식이 올바릅니다.",
        placeholder: "이메일을 입력해주세요.",
        type: "email",
        value: email,
        onChange: onChangeEmail,
        validation: emailValidation,
        label: "이메일"
    };

    const passwordInputProps = {
        testId: "password-input",
        validationMessage: passwordValidationMessage,
        validationTrueMessage: "사용 가능한 비밀번호입니다.",
        placeholder: "비밀번호를 입력해주세요.",
        type: "password",
        value: password,
        onChange: onChangePassword,
        validation: passwordValidation,
        label: "비밀번호"
    };

    const handleEmailValidationFalse = useCallback(() => {
        setEmailValidation(false);
        setEmailValidationMessage("이미 존재하는 이메일입니다.");
    }, [setEmailValidation, setEmailValidationMessage]);

    const onSubmit = (e: FormEvent<HTMLFormElement>) => {
        e.preventDefault();
        if (emailValidation && passwordValidation) {
            if (window.confirm("회원가입을 진행하시겠습니까?")) {
                postSignup({email, password}).then((res) => {
                    if (res) {
                        navigate("/signin");
                    }
                }).catch((err) => {
                    if (err.response.status === 400) {
                        alert("이미 존재하는 이메일입니다.");
                        handleEmailValidationFalse();
                        const inputRef = document.getElementById("validation-input");
                        const emailInput = inputRef?.dataset.testid === "email-input" ? inputRef : null;
                        if (emailInput) {
                            emailInput.focus();
                        }
                    }
                });
            }
        }
    };

    return (
        <StyledFormControl onSubmit={onSubmit}>
            <Typography>
                회원가입
            </Typography>
            <StyledInputBox>
                <ValidationInput {...emailInputProps}/>
            </StyledInputBox>
            <StyledInputBox>
                <ValidationInput {...passwordInputProps}/>
            </StyledInputBox>
            <StyledSignupButton data-textid={"signup-button"} disabled={!emailValidation || !passwordValidation}
                                variant={"contained"} type="submit">
                회원가입
            </StyledSignupButton>
        </StyledFormControl>
    );
};

export default SignUp;
