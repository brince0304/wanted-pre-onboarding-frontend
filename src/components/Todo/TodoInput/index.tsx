import {Button} from "@mui/material";
import TextField from "@mui/material/TextField";
import React, {useRef} from "react";
import {useTokenState} from "../../../context";
import {createTodo} from "../../../apis";
import styled from "@emotion/styled";
import {useFormControl} from "../../../hooks/useFormControl";

const StyledFormControl = styled.form`
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
  align-items: center;
  width: 100%;
  height: 70px;
  background-color: #f5f5f5;
  border-radius: 10px;
  border-bottom: 1px solid #e0e0e0;
  box-shadow: 0 1px 0 0 #e0e0e0;
  padding: 0 20px;
`;

interface TodoInputProps {
    todo: string;
}

const TodoInput = (props: { getTodoList: () => void }) => {
    const regex = /^.{1,}$/;
    const [onChange, value, setValue, validation, setValidation] = useFormControl({ regex });
    const tokenState = useTokenState();
    const isAuth = tokenState.accessToken !== null;
    const ref = useRef<HTMLInputElement>(null);

    const onSubmit = (e: React.FormEvent<HTMLFormElement>) => {
        e.preventDefault();
        if (isAuth && validation) {
            createTodo({ todo: value })
                .then((res) => {
                    props.getTodoList();
                    setValue("");
                    setValidation(false);
                    ref.current?.focus();
                })
                .catch((err) => {
                    console.log(err);
                });
        }
    };

    return (
        <StyledFormControl onSubmit={onSubmit}>
            <TextField
                inputProps={{
                    "data-testid": "new-todo-input",
                }}
                ref={ref}
                variant="standard"
                label="TODO"
                placeholder="할일을 적어보세요!"
                onChange={onChange}
                value={value}
                sx={{ width: "100%" }}
            />
            <Button data-testid="new-todo-add-button" type="submit" disabled={!validation}>
                추가
            </Button>
        </StyledFormControl>
    );
};

export default TodoInput;
