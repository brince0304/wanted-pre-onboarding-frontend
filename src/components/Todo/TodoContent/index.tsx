import {TodoProperties, TodoPropertiesChild} from "../../../interfaces/TodoProperties";
import {
    Box,
    Button,
    Checkbox, FormControl,
    Grow,
    IconButton,
    ListItem, TextField,
    Tooltip,
    Typography,
} from "@mui/material";
import styled from "@emotion/styled";
import {ChangeEvent, useCallback, useEffect, useState} from "react";
import EditIcon from "@mui/icons-material/Edit";
import DeleteIcon from "@mui/icons-material/Delete";
import { useRef } from "react";
import { useFormControl } from "../../../hooks/useFormControl";
import {UpdateTodoData} from "../../../apis/data";
import {deleteTodo, getTodos, updateTodo} from "../../../apis";
import CheckCircleIcon from '@mui/icons-material/CheckCircle';
import CancelIcon from '@mui/icons-material/Cancel';
import {useTokenState} from "../../../context";

const StyledListItem = styled(ListItem)`
  display: flex;
  @keyframes marquee {
    0% { transform: translateX(-10%); }
    100% { transform: translateX(-60%); }
  }
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  width: 100%;
  gap: 10px;
`;

const StyledContentBox = styled(Box)`
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
  align-items: center;
  width: 400px;
  gap: 10px;
`;

const StyledTodoFormControl = styled(FormControl)`
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
  align-items: center;
  width: 400px;
  gap: 10px;
`;

const StyledTodoContentBox = styled(Box)`
    display: block;
  white-space: nowrap;
  overflow: hidden;
    width: 260px;
`;

const OverFlowTitle = styled(Typography)`
  display: inline-block;
  animation: marquee 10s linear infinite;
`

const NonOverFlowTitle = styled(Typography)`
    display: inline-block;
`

const TodoContent = (props: {data:TodoPropertiesChild, getTodoList:()=>void}) => {
    const regex = /^.{1,}$/;
    const [onChange,value,setValue,validation,setValidation] = useFormControl({regex: regex,
    initialValue: props.data.todo,});
    const todoContentRef = useRef<HTMLDivElement>(null)
    const [isChecked, setIsChecked] = useState<boolean>(props.data.isCompleted);
    const [content, setContent] = useState<string>(props.data.todo);
    const [isEdit, setIsEdit] = useState<boolean>(false);
    const [isOverFlow, setIsOverFlow] = useState<boolean>(false);
    const isAuth = useTokenState().accessToken!==null;
    const handleChange = (event: ChangeEvent<HTMLInputElement>) => {
        if(!isAuth) return;
        handleUpdateTodo({
            id: props.data.id.toString(),
            isCompleted: event.target.checked,
            todo: content
        })
    }

    const handleUpdateTodo = (data :UpdateTodoData) => {
        if(!isAuth) return;
       updateTodo(data).then((res) => {
           setContent(res.todo);
           setIsChecked(res.isCompleted);
           props.getTodoList();
       }).catch((err) => {
           console.log(err);
       })
    }

    const handleEditClick = useCallback(() => {
        setIsEdit(true);
    }, []);

    const handleCancelClick = () => {
        setIsEdit(false);
        setValue(content);
    }

    const handleUpdateClick = () => {
        if(value===content){
            setIsEdit(false);
            return;
        }
        if(validation && window.confirm("수정하시겠습니까?")) {
            setIsEdit(false);
            handleUpdateTodo({
                id: props.data.id.toString(),
                isCompleted: isChecked,
                todo: value
            })
        }
    }

    const handleDeleteClick = () => {
        if(!isAuth) return;
        if(window.confirm("삭제하시겠습니까?")) {
            deleteTodo(props.data.id.toString()).then((res) => {
                if(res){
                    setContent("");
                    setIsChecked(false);
                }
            }).catch((err) => {
                console.log(err);
            });
        }
        props.getTodoList();

    }

    useEffect(() => {
        const contentBox = todoContentRef.current;
        if (contentBox) {
            const isOverflowed = contentBox.offsetWidth < contentBox.scrollWidth;
            if(isOverflowed) {
                setIsOverFlow(true)
            }else{
                setIsOverFlow(false)
            }
        }
    }, [content])

    if(content!==""){
    return (
    <StyledListItem>
        <StyledContentBox>
            <Checkbox
                checked={isChecked}
                onChange={handleChange}
                inputProps={{"aria-label": "todo-checkbox"}}
            />
            {!isEdit &&
                <StyledTodoContentBox ref={todoContentRef}>
                    {isOverFlow ? <OverFlowTitle variant={"h6"}>{content}</OverFlowTitle> :
                        <NonOverFlowTitle variant={"h6"}>{content}</NonOverFlowTitle>}
                </StyledTodoContentBox>}
            {isEdit &&
                <StyledTodoContentBox>
                    <TextField variant={"standard"} inputProps={{"data-testid": "modify-input"}} defaultValue={value}
                               value={value} onChange={onChange}
                               fullWidth/>
                </StyledTodoContentBox>
            }

        </StyledContentBox>
        {!isEdit &&
            <StyledContentBox>
                <Tooltip title={"수정"} arrow placement={"top"}>
                    <IconButton data-testid={"modify-button"} onClick={handleEditClick}>
                        <EditIcon/>
                    </IconButton>
                </Tooltip>
                <Tooltip title={"삭제"} arrow placement={"top"}>
                    <IconButton data-testid={"delete-button"} onClick={handleDeleteClick}>
                        <DeleteIcon/>
                    </IconButton>
                </Tooltip>
            </StyledContentBox>}
        {isEdit &&
            <StyledContentBox>
                <Tooltip title={"제출"} arrow placement={"top"}>
                    <IconButton data-testid={"submit-button"} onClick={handleUpdateClick} disabled={!validation}>
                        <CheckCircleIcon/>
                    </IconButton>
                </Tooltip>
                <Tooltip title={"취소"} arrow placement={"top"}>
                    <IconButton data-testid={"cancel-button"} onClick={handleCancelClick}>
                        <CancelIcon/>
                    </IconButton>
                </Tooltip>
            </StyledContentBox>}


    </StyledListItem>
    );}else{
        return null;
    }
};

export default TodoContent;
