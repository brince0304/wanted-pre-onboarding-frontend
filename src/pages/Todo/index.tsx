import {Box, List, Typography} from "@mui/material";
import styled from "@emotion/styled";
import {useEffect, useState} from "react";
import {TodoProperties} from "../../interfaces/TodoProperties";
import {getTodos} from "../../apis";
import TodoInput from "../../components/Todo/TodoInput";
import {useTokenState} from "../../context";
import TodoContent from "../../components/Todo/TodoContent";
import {useNavigate} from "react-router";

const StyledBox = styled(Box)`
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  align-items: center;
  width: 100%;
  gap: 20px;
`;


const StyledTodoList = styled(List)`
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  align-items: center;
  width: 450px;
  background-color: #f5f5f5;
  border-radius: 10px;
  border-bottom: 1px solid #e0e0e0;
  box-shadow: 0 1px 0 0 #e0e0e0;
`;


const Todo = () => {
    const [isLoading,setIsLoading] = useState<boolean>(true);
    const [data, setData] = useState<TodoProperties>([]);
    const tokenState = useTokenState();
    const isAuth = tokenState.accessToken !== null;
    const navigate = useNavigate();
    useEffect(()=>{
        if(isAuth) {
            getTodos().then((res) => {
                setData(res);
                setIsLoading(false);
            })
        }else{
            window.alert("로그인이 필요합니다.");
            navigate("/signin");
        }
    },[])

    const getTodoList = () => {
        if(!isAuth){
            window.alert("로그인이 필요합니다.");
            navigate("/signin");
            return;
        }
        getTodos().then((res)=>{
            setData(res);
        })
    }
    return (
        <StyledBox>
            <TodoInput getTodoList={getTodoList}/>
            <StyledTodoList>
                {isLoading ? <Typography variant={"h6"}>로딩중..</Typography> : (
                    data && data.length > 0 ?
                        data.map((todo)=>{
                            return <TodoContent key={todo.id} data={todo} getTodoList={getTodoList} />
                        }) : <Typography variant={"h6"}>할 일이 없습니다.</Typography>
                    )}
            </StyledTodoList>
        </StyledBox>
    )
}

export default Todo
