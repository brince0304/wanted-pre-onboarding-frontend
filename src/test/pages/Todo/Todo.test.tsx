import {TodoPropertiesChild} from "../../../interfaces/TodoProperties";
import SignUp from "../../../pages/SignUp";
import {act, fireEvent, render, screen, waitFor} from "@testing-library/react";
import Todo from "../../../pages/Todo";
import React, {ReactNode} from "react";
import  {TokenProvider} from "../../../context";
import {BrowserRouter} from "react-router-dom";
import * as router from "react-router-dom";


const apiMethods = require("../../../apis/index.ts");
const mockedUsedNavigate = jest.fn();

jest.mock("react-router", () => ({
    ...jest.requireActual("react-router"),
    useNavigate: () => mockedUsedNavigate
}));

beforeEach(() => {
    mockedUsedNavigate.mockReset();
});

afterEach(() => {
    jest.clearAllMocks();
    jest.restoreAllMocks();
});


function customRender(children: ReactNode) {
    return render(
        <TokenProvider>
            <BrowserRouter>
                {children}
            </BrowserRouter>
        </TokenProvider>
    );
}

const todoList: TodoPropertiesChild[] = [
    {
        id: 1,
        todo: "투두입니다.",
        isCompleted: false,
        userId: 1
    }];

const updatedTodoList: TodoPropertiesChild[] = [
    {
        id: 1,
        todo: "수정된 투두입니다.",
        isCompleted: false,
        userId: 1
    }];

const giveTokenProvider = () => {
    const context = require("../../../context/index.tsx");
    context.useTokenState = jest.fn().mockReturnValue("token");
};

const giveEmptyArray = () => {
    giveTokenProvider();
    const getTodos = jest.spyOn(apiMethods, "getTodos").mockResolvedValue([]);
};

const giveTodoList = () => {
    giveTokenProvider();
    // 목록 - 수정 - 삭제
    const getTodos = jest.spyOn(apiMethods, "getTodos").mockResolvedValueOnce(todoList)
        .mockResolvedValueOnce(updatedTodoList);
};

function giveEmptyTokenProvider() {
    const context = require("../../../context/index.tsx");
    context.useTokenState = jest.fn().mockReturnValue({accessToken:null});
}

describe("<Todo/>", () => {
    it("matches snapshot", () => {
        customRender(<Todo/>);
        const utils = customRender(<Todo/>);
        expect(utils.container).toMatchSnapshot();
    });
    it("토큰이 없으면 로그인 페이지로 이동", async () => {
        giveEmptyTokenProvider();
        customRender(<Todo/>);
        expect(mockedUsedNavigate).toHaveBeenCalledWith("/signin");
    });
    it("처음 투두를 랜더링할 시에 로딩중 문구 출력, 로딩 후 할일이 없습니다 출력", async () => {
        giveEmptyArray();
        customRender(<Todo/>);
        expect(screen.getByText("로딩중..")).toBeInTheDocument();
        await waitFor(() => {
            expect(screen.queryByText("로딩중..")).toBeNull();
        });
    });
    it("투두가 존재하면 투두 리스트 출력", async () => {
        giveTodoList();
        customRender(<Todo/>);
        await waitFor(() => {
            expect(screen.getAllByText("투두입니다.")).toHaveLength(1);
        });
    });
    it("투두가 존재하지 않으면 할일이 없습니다 출력", async () => {
        giveEmptyArray();
        customRender(<Todo/>);
        await waitFor(() => {
            expect(screen.getByText("할 일이 없습니다.")).toBeInTheDocument();
        });
    });
    it("투두 입력 안되어있을 때 버튼 disabled", () => {
        giveEmptyArray();
        customRender(<Todo/>);
        const inputButton = screen.getByTestId("new-todo-add-button");
        const todoInput = screen.getByTestId("new-todo-input");
        fireEvent.change(todoInput, {target: {value: ""}});
        expect(inputButton).toBeDisabled();
    });
    it("투두 조건에 맞게 입력이 되어있을 때 버튼 disabled 해제", () => {
        giveEmptyArray();
        customRender(<Todo/>);
        const inputButton = screen.getByTestId("new-todo-add-button");
        const todoInput = screen.getByTestId("new-todo-input");
        fireEvent.change(todoInput, {target: {value: "투두입니다."}});
        expect(inputButton).not.toBeDisabled();
    });
    it("투두 입력 후 추가 버튼 클릭시 createTodo 호출 후 value 값 초기화 ", async () => {
        giveEmptyArray();
        const createTodo = jest.spyOn(apiMethods, "createTodo").mockResolvedValue({});
        customRender(<Todo/>);
        const inputButton = screen.getByTestId("new-todo-add-button");
        const todoInput = screen.getByTestId("new-todo-input");
        fireEvent.change(todoInput, {target: {value: "투두입니다."}});
        fireEvent.click(inputButton);
        await waitFor(() => {
            expect(createTodo).toHaveBeenCalled();
        });
        await waitFor(()=> {
            expect(createTodo).toHaveBeenCalledWith({todo: "투두입니다."});
        })
        await waitFor(() => {
            expect(todoInput).toHaveValue("");
        });
    });
    it("수정 버튼 클릭했을시 수정 컴포넌트 랜더링", async () => {
        giveTodoList();
        customRender(<Todo/>);
        await waitFor(() => {
            expect(screen.getByTestId("modify-button")).toBeInTheDocument();
        });
        fireEvent.click(screen.getByTestId("modify-button"));
        await waitFor(() => {
            expect(screen.getByTestId("modify-input")).toBeInTheDocument();
        });
    });
    it("수정 버튼 클릭했을시 수정 컴포넌트 랜더링 후 수정 버튼 사라짐", async () => {
        giveTodoList();
        customRender(<Todo/>);
        await waitFor(() => {
            const modifyButton1 = screen.getByTestId("modify-button") as HTMLButtonElement;
            expect(modifyButton1).toBeInTheDocument();
        });
        fireEvent.click(screen.getByTestId("modify-button"));
        await waitFor(() => {
            expect(screen.queryByTestId("modify-button")).not.toBeInTheDocument();
        });
    });
    it("투두가 공백일시에 제출 버튼 비활성화", async () => {
        giveTodoList();
        customRender(<Todo/>);
        await waitFor(() => {
            const modifyButton1 = screen.getByTestId("modify-button") as HTMLButtonElement;
            expect(modifyButton1).toBeInTheDocument();
        });
        fireEvent.click(screen.getByTestId("modify-button"));
        await waitFor(() => {
            expect(screen.getByTestId("modify-input")).toBeInTheDocument();
        });
        const modifyInput = screen.getByTestId("modify-input");
        const modifyButton = screen.getByTestId("submit-button");
        fireEvent.change(modifyInput, {target: {value: ""}});
        expect(modifyButton).toBeDisabled();
    });
    it("취소 버튼 클릭했을시 수정 컴포넌트 사라지고 투두 출력", async () => {
        giveTodoList();
        customRender(<Todo/>);
        await waitFor(() => {
            const modifyButton1 = screen.getByTestId("modify-button") as HTMLButtonElement;
            expect(modifyButton1).toBeInTheDocument();
        });
        fireEvent.click(screen.getByTestId("modify-button"));
        await waitFor(() => {
            expect(screen.getByTestId("cancel-button")).toBeInTheDocument();
        });
        fireEvent.click(screen.getByTestId("cancel-button"));
        await waitFor(() => {
            expect(screen.getByText("투두입니다.")).toBeInTheDocument();
        });
    });
    it("수정내용 입력하고 수정버튼 클릭하면 api 호출 후 반영", async () => {
        giveTodoList();
        jest.spyOn(window,'confirm').mockImplementation(() => true);
        const updateTodo = jest.spyOn(apiMethods, "updateTodo").mockResolvedValue({});
        customRender(<Todo/>);
        await waitFor(() => {
            const modifyButton1 = screen.getByTestId("modify-button") as HTMLButtonElement;
            expect(modifyButton1).toBeInTheDocument();
        });
        fireEvent.click(screen.getByTestId("modify-button"));
        const submitButton = screen.getByTestId("submit-button");
        const modifyInput = screen.getByTestId("modify-input");
        expect(submitButton).toBeInTheDocument();
        expect(modifyInput).toBeInTheDocument();
        fireEvent.change(modifyInput, {target: {value: "수정된 투두입니다."}});
        fireEvent.click(submitButton);
        await waitFor(() => {
            expect(updateTodo).toHaveBeenCalled();
        });
        await waitFor(() => {
            expect(screen.getByText("수정된 투두입니다.")).toBeInTheDocument();
        });
    });
    it("삭제 버튼 클릭하면 확인 후 빈 배열 반환", async () => {
        giveTokenProvider();
        // 목록 - 수정 - 삭제
        jest.spyOn(apiMethods, "getTodos").mockResolvedValueOnce(todoList)
            .mockResolvedValueOnce([]);
        jest.spyOn(window,'confirm').mockImplementation(() => true);
        const deleteTodo = jest.spyOn(apiMethods, "deleteTodo").mockResolvedValue({response:{
            data: [],
            status: 204
            }});
        customRender(<Todo/>);
        await waitFor(() => {
            const deleteButton = screen.getByTestId("delete-button") as HTMLButtonElement;
            expect(deleteButton).toBeInTheDocument();
        });
        fireEvent.click(screen.getByTestId("delete-button"));
        await waitFor(() => {
            expect(deleteTodo).toHaveBeenCalled();
        });
        await waitFor(() => {
            expect(screen.getByText("할 일이 없습니다.")).toBeInTheDocument();
        });
    });
    it("토큰 없는 상황에서 삭제 시도하면 로그인 페이지로 이동", async () => {
        giveEmptyTokenProvider();
        const getTodos = jest.spyOn(apiMethods, "getTodos").mockResolvedValueOnce(todoList)
            .mockResolvedValueOnce(updatedTodoList);
        jest.spyOn(window,'confirm').mockImplementation(() => true);
        jest.spyOn(window,'alert').mockImplementation(() => {});
        const deleteTodo = jest.spyOn(apiMethods, "deleteTodo").mockRejectedValue({response:{
                data: [],
                status: 401
            }});
        customRender(<Todo/>);
        await expect(mockedUsedNavigate).toHaveBeenCalledWith("/signin");
    });
    it("토큰 없는 상황에서 수정 시도하면 로그인 페이지로 이동", async () => {
        giveEmptyTokenProvider();
        const getTodos = jest.spyOn(apiMethods, "getTodos").mockResolvedValueOnce(todoList)
            .mockResolvedValueOnce(updatedTodoList);
        const updateTodo = jest.spyOn(apiMethods, "updateTodo").mockRejectedValue({response:{
                data: [],
                status: 401
            }});
        jest.spyOn(window,'confirm').mockImplementation(() => true);
        jest.spyOn(window,'alert').mockImplementation(() => {});
        customRender(<Todo/>);
        await expect(mockedUsedNavigate).toHaveBeenCalledWith("/signin");
    });
    it("토큰 없는 상황에서 추가 시도하면 로그인 페이지로 이동", async () => {
        giveEmptyTokenProvider();
        const getTodos = jest.spyOn(apiMethods, "getTodos").mockResolvedValueOnce(todoList)
            .mockResolvedValueOnce(updatedTodoList);
        const createTodo = jest.spyOn(apiMethods, "createTodo").mockRejectedValue({response:{
                data: [],
                status: 401
            }});
        jest.spyOn(window,'confirm').mockImplementation(() => true);
        jest.spyOn(window,'alert').mockImplementation(() => {});
        customRender(<Todo/>);
        await expect(mockedUsedNavigate).toHaveBeenCalledWith("/signin");
    });
});