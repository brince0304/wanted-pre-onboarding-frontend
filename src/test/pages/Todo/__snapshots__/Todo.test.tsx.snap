// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Todo/> matches snapshot 1`] = `
<div>
  <div
    class="MuiBox-root css-bvbvw"
  >
    <form
      class="css-1wg1k7o"
    >
      <div
        class="MuiFormControl-root MuiTextField-root css-136h9lv-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard css-aqpgxn-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for=":r1:"
          id=":r1:-label"
        >
          TODO
        </label>
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl css-v4u5dn-MuiInputBase-root-MuiInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
            data-testid="new-todo-input"
            id=":r1:"
            placeholder="할일을 적어보세요!"
            type="text"
            value=""
          />
        </div>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium Mui-disabled MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
        data-testid="new-todo-add-button"
        disabled=""
        tabindex="-1"
        type="submit"
      >
        추가
      </button>
    </form>
    <ul
      class="MuiList-root MuiList-padding css-1bkqi2j-MuiList-root"
    >
      <h6
        class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
      >
        로딩중..
      </h6>
    </ul>
  </div>
</div>
`;
